# 概要

# 予約語
- ECMAScriptで定義されている
    - 変数名や関数名には使用できない
- ES2020の場合
    - await, break, case, catch, class, const, continue, debugger, default, delete, do, else, enum, export, extends, false, finally, for, function, if, import, in, instanceof, new, null, return, super, switch, this, throw, true, try, typeof, var, void, while, with, yield
- strict mode
    - let, static, implements, interface, package, private, protected, public

In [8]:
let if = 10;　//予約語は変数名には使えない

SyntaxError: Unexpected token 'if'

# データ型
- プリミティブ型：単一の値を持つ
    - 数値型
    - 長数型 (ES2020~)
    - 文字列
    - 理論値
    - undifined
    - null
    - シンボル　(ES2015~)：作成するたびに新しいユニークな値ができる
- オブジェクト型：プリミティブ型以外
    - オブジェクト
    - 配列
    - 関数
    - 日時
    - 正規表記
    - JSON
    - ラッパーオブジェクト：undefinedとnullを除いたプリミティブ型に対して用意されている
        - プリミティブ型に対してラッパーで定義されているメソッドを使うと自動的に変換される

## プリミティブ型

In [1]:
console.log(typeof(10)); //number
console.log(typeof(10n)); //longint
console.log(typeof('Hello')); //string
console.log(typeof(true)); //boolean

number
bigint
string
boolean


In [14]:
var num
console.log(typeof(num)); //object
console.log(typeof(undefined)) //undifined

undefined
undefined


In [4]:
var num = null
console.log(num)
console.log(typeof(num));
console.log(typeof(null))

null
object
object


In [7]:
let s1 = Symbol();
let s2 = Symbol();
console.log(typeof(s1))
console.log(s1 === s2);

symbol
false


### 数値型

- 浮動小数点数

In [22]:
console.log(3.14)
console.log(3.14e3)

3.14
3140


- 数値型では2進数や8進数を扱える

In [9]:
console.log(0b10)
console.log(0o10)
console.log(0x10)
console.log(010) //8進数になるがstrictモードではエラー

2
8
16
8


- NaNやInfinityは数値型

In [17]:
console.log(NaN);
console.log(typeof(NaN));
console.log(Infinity);
console.log(typeof(Infinity));

NaN
number
Infinity
number


In [19]:
console.log(NaN + 1);
console.log(Infinity + 1);

NaN
Infinity


### 長数型

In [41]:
console.log(0b10n)
console.log(0o10n)
console.log(0x10n)

2n
8n
16n


### 論理値

- falseとみなされる値
    - false, 0, -0, 0n, "", null, undefined, NaN

In [21]:
if (0) {
  console.log("true");
} else {
  console.log("false");
}

false


## オブジェクト型

In [1]:
const myArray = new Array(1, 2, 3, 4, 5);

function f(){};

const currentDate = new Date();

const regexPattern = /^([a-zA-Z0-9_-]+)$/;

const jsonData = {
  name: "John",
  age: 30,
  hobbies: ["reading", "coding", "hiking"],
  isEmployed: true
};

In [2]:
console.log(typeof(myArray))
console.log(typeof(f))
console.log(typeof(currentDate))
console.log(typeof(regexPattern))
console.log(typeof(jsonData));

object
function
object
object
object


In [3]:
// 詳細な情報
console.log(Object.prototype.toString.call(myArray));
console.log(Object.prototype.toString.call(f));
console.log(Object.prototype.toString.call(currentDate));
console.log(Object.prototype.toString.call(regexPattern));
console.log(Object.prototype.toString.call(jsonData));

[object Array]
[object Function]
[object Date]
[object RegExp]
[object Object]


### オブジェクト

In [2]:
let user = {
    name:"太郎",
    age:22,
    address:"東京",
    hobby:["サッカー", "カラオケ"],
    getAge:function(){
        return this.age;
    }
}

In [4]:
console.log(user.name);
console.log(user['name']);
console.log(user.getAge());

太郎
太郎
22


### Numberオブジェクト

In [2]:
let num = 10;
let numObj = new Number(num);
let numNotObj = Number(num) //new　をつけないとプリミティブ型となる
console.log(num);
console.log(numObj);
console.log(numNotObj);
console.log(typeof(numObj));
console.log(num == numObj);
console.log(num === numObj);

10
[Number: 10]
10
object
true
false


- 静的プロパティ

In [5]:
console.log(Number.MAX_SAFE_INTEGER);
console.log(Number.MIN_SAFE_INTEGER);
console.log(Number.MAX_VALUE);
console.log(Number.MIN_VALUE);
console.log(Number.NaN);
console.log(Number.POSITIVE_INFINITY);
console.log(Number.NEGATIVE_INFINITY);

9007199254740991
-9007199254740991
1.7976931348623157e+308
5e-324
NaN
Infinity
-Infinity


### 型変換

- toString
    - 基数を指定することができる

In [1]:
let num1 = 10
console.log(num1.toString(), typeof(num1.toString()))

10 string


In [7]:
// console.log(1.toString)はエラー
console.log(1 .toString());
console.log((1).toString());

1
1


In [9]:
console.log(20 .toString(2));
console.log(20 .toString(8));
console.log(20 .toString(20));

10100
24
10


- toExponential
    - 小数点以下の桁数を指定できる（2への丸め）

In [8]:
console.log(1 .toExponential());

1e+0


In [11]:
console.log(1.4142 .toExponential(2));
console.log(1.4142 .toExponential(3));

1.41e+0
3.414e+0


- toFixed
    - 小数点以下の桁数を指定できる

In [14]:
console.log(1.4142e1 .toFixed());
console.log(1.4142e1 .toFixed(2));

14
14.14


- toPrecision
    - 有効桁数を指定できる

In [16]:
console.log(1.4142e1 .toPrecision());
console.log(1.4142e1 .toPrecision(2));
console.log(1.4142e1 .toPrecision(7));

14.142
14
14.14200


- toLocaleString

In [2]:
console.log(1.4142e4 .toLocaleString());
console.log(1.4142e4 .toLocaleString('ja-JP'));
console.log(1.4142e4 .toLocaleString('de-DE'));
console.log(1.4142e4 .toLocaleString('ru-RU'));

14,142
14,142
14.142
14 142


- Number.parseInt
    - 基数を指定して文字列を数値に変換できる

In [21]:
console.log(Number.parseInt('10'))
console.log(Number.parseInt('0.1'))
console.log(Number.parseInt('0.9'))
console.log(Number.parseInt('10e3'))
console.log(Number.parseInt('10', 2))
console.log(Number.parseInt('0x10'))

10
0
0
10
2
16


- parseFloat

In [25]:
console.log(Number.parseFloat('10'))
console.log(Number.parseFloat('0.1'))
console.log(Number.parseFloat('10e3'))
console.log(Number.parseFloat('10abc'))
console.log(Number.parseFloat('a10'))
console.log(Number.parseFloat('0.101', 2))
console.log(Number.parseFloat('10', 2))
console.log(Number.parseFloat('0x10'))

10
0.1
10000
10
NaN
0.101
10
0


- Nyumber.isNaN
    - 引数が数値型でない場合全てfalse

In [28]:
console.log(Number.isNaN(NaN))
console.log(Number.isNaN(0/0))
console.log(Number.isNaN(undefined))
console.log(Number.isNaN(Number(undefined)))

true
true
false
true


- isNaN
    - 引数が数値型でない場合数値型に変換

In [31]:
console.log(isNaN('Hello'));
console.log(Number.isNaN('Hello'));
console.log(isNaN(undefined));

true
false
true


- Number.isFinite

In [32]:
console.log(Number.isFinite(10));
console.log(Number.isFinite(Infinity));
console.log(Number.isFinite(-Infinity));

true
false
false


- isFinite

In [14]:
console.log(Number('10'))
console.log(Number('0.1'))
console.log(Number('abc'))
console.log(Number(10n))
console.log(Number([10]))
console.log(Number([10, 20]))
console.log(Number(undefined))
console.log(Number(null))
console.log(Number(NaN))

10
0.1
NaN
10
10
NaN
NaN
0
NaN


- Number.isInteger

In [36]:
console.log(Number.isInteger(10));
console.log(Number.isInteger(10.0));
console.log(Number.isInteger(1.0e1));
console.log(Number.isInteger('10'));
console.log(Number.isInteger(NaN));
console.log(Number.isInteger(Infinity));

true
true
true
false
false
false


- Number.isSafeInteger
    - 正確に IEEE-754 倍精度数として表すことができ、別の値を IEEE-754 倍精度数として表した値と一致することがない値
    - 具体的には -(253 - 1) 以上 253 - 1 以下の整数値です (-9007199254740991 以上 9007199254740991 以下の整数値です)

In [43]:
console.log(Number.MAX_SAFE_INTEGER);
console.log(Number.MAX_SAFE_INTEGER+1);
console.log(Number.MAX_SAFE_INTEGER+2);

9007199254740991
9007199254740992
9007199254740992


In [40]:
console.log(Number.isSafeInteger(10));
console.log(Number.isSafeInteger(Math.pow(2,53)));

true
false


- +演算子を使った場合、文字列が含まれると自動的に文字列に変換される
- ただしシンボルは例外

In [12]:
console.log('1' + 2);
console.log('1' + 2 + 3);
console.log('1' + 2 + 3n);
console.log('1' + 2 + -3);
console.log('1' + 2 + true + [4, 5] + null + undefined);

12
123
123
12-3
12true4,5nullundefined


- -演算子や\*演算子を使った場合、数値以外の値が数値になる
- シンボルは例外

In [4]:
console.log('1' - 2);
console.log('1' - 2 - 3);
console.log('1' - 2 - true - [4, 5] - null - undefined);

-1
-4
NaN


## strictモード
- 推奨されない記述方法をエラーにすることができる
- プログラムや関数の初めに'use strict'を記述

In [1]:
function f(){
    'use strict';
    let name;
    nama = 'Taro'; //未定義の変数への代入なので（本当は）エラーとなる
}

In [2]:
function f(){
    let name;
    nama = 'Taro';
}

In [3]:
// withの使用は分かりにくく、速度的にも問題があるためstrictモードではできない
with(Math){
  console.log(min(10,20,30));
  console.log(max(10,20,30));
  console.log(pow(4, 2));
}

10
30
16
